components/director/internal/domain/formationconstraint/constraintEngine.go

```diff
@@ -0,0 +1,130 @@
+		operatorInputConstructors: map[OperatorName]OperatorInputConstructor{IsNotAssignedToAnyFormationOfTypeOperator: NewIsNotAssignedToAnyFormationOfTypeInput},
+	}
+	c.operators = map[OperatorName]OperatorFunc{IsNotAssignedToAnyFormationOfTypeOperator: c.IsNotAssignedToAnyFormationOfType}
+	return c
+}
+
+// EnforceConstraints finds all the applicable constraints based on JoinPointLocation and JoinPointDetails. Checks for each constraint if it is satisfied.
+// If any constraint is not satisfied this information is stored and the engine proceeds with enforcing the next constraint if such exists. In the end if
+// any constraint was not satisfied an error is returned.
+func (e *ConstraintEngine) EnforceConstraints(ctx context.Context, location formationconstraint.JoinPointLocation, details formationconstraint.JoinPointDetails, formationTemplateID string) error {
+	matchingDetails := details.GetMatchingDetails()
+	log.C(ctx).Infof("Enforcing constraints for target operation %q, constraint type %q, resource type %q and resource subtype %q", location.OperationName, location.ConstraintType, matchingDetails.ResourceType, matchingDetails.ResourceSubtype)
+
+	constraints, err := e.constraintSvc.ListMatchingConstraints(ctx, formationTemplateID, location, matchingDetails)
+	if err != nil {
+		return errors.Wrapf(err, "While listing matching constraints for target operation %q, constraint type %q, resource type %q and resource subtype %q", location.OperationName, location.ConstraintType, matchingDetails.ResourceType, matchingDetails.ResourceSubtype)
+	}
+
+	matchedConstraintsNames := make([]string, 0, len(constraints))
+	for _, c := range constraints {
+		matchedConstraintsNames = append(matchedConstraintsNames, c.Name)
+	}
+
+	log.C(ctx).Infof("Matched constraints: %v", matchedConstraintsNames)
+
+	var errs *multierror.Error
+	for _, mc := range constraints {
+		operator, ok := e.operators[OperatorName(mc.Operator)]
+		if !ok {
+			errs = multierror.Append(errs, ConstraintError{
+				ConstraintName: mc.Name,
+				Reason:         fmt.Sprintf("Operator %q not found", mc.Operator),
+			})
+			continue
+		}
+
+		operatorInputConstructor, ok := e.operatorInputConstructors[OperatorName(mc.Operator)]
+		if !ok {
+			errs = multierror.Append(errs, ConstraintError{
+				ConstraintName: mc.Name,
+				Reason:         fmt.Sprintf("Operator input constructor for operator %q not found", mc.Operator),
+			})
+			continue
+		}
+
+		operatorInput := operatorInputConstructor()
+		if err := formationconstraint.ParseInputTemplate(mc.InputTemplate, details, operatorInput); err != nil {
+			log.C(ctx).Errorf("An error occurred while parsing input template for formation constraint %q: %s", mc.Name, err.Error())
+			errs = multierror.Append(errs, ConstraintError{
+				ConstraintName: mc.Name,
+				Reason:         fmt.Sprintf("Failed to parse operator input template for operator %q", mc.Operator),
+			})
+			continue
+		}
+
+		operatorResult, err := operator(ctx, operatorInput)
```

Comment:
If the operator has to retur specific type error that wont be a problem as it is still error. The name operator may be a bit misleading - the idea behind the operators is to check if a some requirements are met. The operator is not expected to do anything else. The 'operators in UI are different from the ones in the backend. The operators in the UI are meant to be more like actuators and are expected to make changes to the views.

Reply:
I asked because the `operatorResult` is bool and we check it in the `if` statement below. And I was wondering if there is a case where we want a specific type instead of bool. If we're using the operators only to check something I guess we can always based on the requirements to return true/false.